### PR TITLE
Add Generalized Harmonic Characteristic Speeds

### PR DESCRIPTION
**Pull request by osheamonn**
## Proposed changes

- Adds a prefix tag CharSpeed. This is useful because for every field, the characteristic speed is just a scalar, so CharSpeed<T> is always a scalar, and it means we can make a variables of characteristic speeds. 
- Adds the characteristic speeds. The function to compute them currently returns a Variables of the speeds. 

### Types of changes:

- [ ] Bugfix
- [ ] New feature

### Component:

- [ ] Code
- [ ] Documentation
- [ ] Build system
- [ ] Continuous integration

### Code review checklist

- [ ] The PR passes all checks, including unit tests, `clang-tidy` and `IWYU`. For
  instructions on how to perform the CI checks locally refer to the [Dev guide
  on the Travis CI](https://spectre-code.org/travis_guide.html).
- [ ] The code is documented and the documentation renders correctly. Run `make doc`
  to generate the documentation locally into `BUILD_DIR/docs/html`. Then open
  `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).

### CODE REVIEW

**Review by kidder — DISMISSED** [review]

File: src/Evolution/Systems/GeneralizedHarmonic/Equations.hpp [hunk 1]
```diff
@@ -41,21 +41,21 @@ namespace GeneralizedHarmonic {
...
-                 GaugeH<Dim>, SpacetimeDerivGaugeH<Dim>, gr::Tags::Lapse<>,
-                 gr::Tags::Shift<Dim>, gr::Tags::InverseSpatialMetric<Dim>,
-                 gr::Tags::InverseSpacetimeMetric<Dim>,
-                 gr::Tags::TraceSpacetimeChristoffelFirstKind<Dim>,
-                 gr::Tags::SpacetimeChristoffelFirstKind<Dim>,
-                 gr::Tags::SpacetimeChristoffelSecondKind<Dim>,
-                 gr::Tags::SpacetimeNormalVector<Dim>,
-                 gr::Tags::SpacetimeNormalOneForm<Dim>>;
+  using argument_tags = tmpl::list<
+      gr::Tags::SpacetimeMetric<Dim>, Tags::Pi<Dim>, Tags::Pi<Dim>,
```
Comment (hrueter) [inline src/Evolution/Systems/GeneralizedHarmonic/Equations.hpp]: Why the change Phi -> Pi ?
Is this a typo?

File: src/Evolution/Systems/GeneralizedHarmonic/Equations.hpp [hunk 2]
```diff
@@ -41,21 +41,21 @@ namespace GeneralizedHarmonic {
...
+      ::Tags::deriv<gr::Tags::SpacetimeMetric<Dim>, tmpl::size_t<Dim>,
+                    Frame::Inertial>,
+      ::Tags::deriv<Tags::Pi<Dim>, tmpl::size_t<Dim>, Frame::Inertial>,
+      ::Tags::deriv<Tags::Pi<Dim>, tmpl::size_t<Dim>, Frame::Inertial>,
```
Comment (hrueter) [inline src/Evolution/Systems/GeneralizedHarmonic/Equations.hpp]: Phi?

File: src/Evolution/Systems/GeneralizedHarmonic/Equations.hpp [hunk 3]
```diff
@@ -111,9 +111,10 @@ template <size_t Dim>
 struct ComputeNormalDotFluxes {
  public:
   using argument_tags =
-      tmpl::list<gr::Tags::SpacetimeMetric<Dim>, Pi<Dim>, Phi<Dim>,
-                 ConstraintGamma1, ConstraintGamma2, gr::Tags::Lapse<>,
-                 gr::Tags::Shift<Dim>, gr::Tags::InverseSpatialMetric<Dim>>;
+      tmpl::list<gr::Tags::SpacetimeMetric<Dim>, Tags::Pi<Dim>, Tags::Pi<Dim>,
```
Comment (hrueter) [inline src/Evolution/Systems/GeneralizedHarmonic/Equations.hpp]: Phi?

File: src/Evolution/Systems/GeneralizedHarmonic/Equations.hpp (hunk 1 above)
Comment (osheamonn) [reply src/Evolution/Systems/GeneralizedHarmonic/Equations.hpp]: Yes that's a typo, thanks.

File: tests/Unit/Evolution/Systems/GeneralizedHarmonic/Test_Characteristics.cpp [hunk 4]
```diff
@@ -0,0 +1,88 @@
...
+// IWYU pragma: no_forward_declare Tensor
+// IWYU pragma: no_forward_declare Variables
+
+namespace {
+template <size_t Dim, typename Frame>
+Scalar<DataVector> upsi_computer(
+    const Scalar<DataVector>& gamma_1, const Scalar<DataVector>& lapse,
+    const tnsr::I<DataVector, Dim, Frame>& shift,
+    const tnsr::i<DataVector, Dim, Frame>& normal) {
+  return get<Tags::CharSpeed<GeneralizedHarmonic::Tags::UPsi<Dim, Frame>>>(
```
Comment (nilsdeppe) [inline tests/Unit/Evolution/Systems/GeneralizedHarmonic/Test_Characteristics.cpp]: If I'm reading these correctly, could you combine them into one passing in the tag to return as a template parameter?

**Review by nilsdeppe — COMMENTED** [review]
One minor thing, I'm happy for you to squash :)

**Review by nilsdeppe — DISMISSED** [review]
@kidder @wthrowe could you please look at this when you have time? Thanks! :)

**Comment by nilsdeppe** [comment]
@osheamonn this needs a rebase because of a conflict.

**Review by wthrowe — DISMISSED** [review]

**Comment by nilsdeppe** [comment]
It looks like every travis build failed...

File: src/Evolution/Systems/GeneralizedHarmonic/Characteristics.hpp [hunk 5]
```diff
@@ -0,0 +1,66 @@
...
+ * v_{0} =& -n_k N^k \\
+ * v_{\pm} =& -n_k N^k \pm N
+ * \f}
+ *
+ * where \f$N, N^k\f$ are the lapse and shift respectively, \f$\gamma_1\f$ is a
+ * constraint damping parameter, and \f$n_k\f$ is the unit normal to the
+ * surface.
+ */
+template <size_t Dim, typename Frame>
+struct ComputeCharacteristicSpeeds : db::ComputeTag {
```
Comment (kidder) [inline src/Evolution/Systems/GeneralizedHarmonic/Characteristics.hpp]: We've decided to postfix `Compute` so simple tags and compute tags show up next to each other in the documentation

**Review by kidder — CHANGES_REQUESTED** [review]

File: tests/Unit/Evolution/Systems/GeneralizedHarmonic/Test_Characteristics.cpp [hunk 6]
```diff
@@ -0,0 +1,73 @@
...
+      compute_speed_with_tag<
+          Tags::CharSpeed<GeneralizedHarmonic::Tags::UMinus<Dim, Frame>>, Dim,
+          Frame>,
+      "TestFunctions", "char_speed_uminus", {{{-10.0, 10.0}}}, used_for_size);
+  pypp::check_with_random_values<1>(
+      compute_speed_with_tag<
+          Tags::CharSpeed<GeneralizedHarmonic::Tags::UPlus<Dim, Frame>>, Dim,
+          Frame>,
+      "TestFunctions", "char_speed_uplus", {{{-10.0, 10.0}}}, used_for_size);
+}
```
Comment (geoffrey4444) [inline tests/Unit/Evolution/Systems/GeneralizedHarmonic/Test_Characteristics.cpp:59]: Besides fuzzy testing, can you also test a simple known case, like Schwarzschild?

**Review by geoffrey4444 — CHANGES_REQUESTED** [review]
Looks good, and I don't want to hold it up. I have one requested change (test an analytic solution as well as fuzzy testing), but if everyone else agrees it's unnecessary, I'm willing to approve anyway.

File: tests/Unit/Evolution/Systems/GeneralizedHarmonic/Test_Characteristics.cpp (hunk 6 above)
Comment (osheamonn) [reply tests/Unit/Evolution/Systems/GeneralizedHarmonic/Test_Characteristics.cpp:59]: That's probably a good idea, is it ok if it's made an issue and I put in another PR for it? This is pretty much ready to merge, and some of this PR is used in the characteristic fields PR.

Comment (kidder) [reply tests/Unit/Evolution/Systems/GeneralizedHarmonic/Test_Characteristics.cpp:59]: I'm fine with making this an issue

**Review by kidder — APPROVED** [review]

**Comment by kidder** [comment]
@geoffrey4444 @nilsdeppe @wthrowe please look at

File: src/Evolution/Systems/GeneralizedHarmonic/Tags.hpp [hunk 7]
```diff
@@ -61,5 +62,42 @@ struct SpacetimeDerivGaugeH : db::SimpleTag {
...
   static std::string name() noexcept { return "SpacetimeDerivGaugeH"; }
 };
+
+// {@
+/// \ingroup GeneralizedHarmonicGroup
+/// \brief Tags corresponding to the characteristic fields of the generalized
+/// system.
+///
+/// \details For details on how these are defined and computed, see
+/// ComputeCharacteristicSpeeds
```
Comment (nilsdeppe) [inline src/Evolution/Systems/GeneralizedHarmonic/Tags.hpp:72]: `CharacteristicSpeedsCompute`

File: src/Evolution/Systems/GeneralizedHarmonic/TagsDeclarations.hpp [hunk 8]
```diff
@@ -21,5 +21,17 @@ template <size_t Dim, typename Frame = Frame::Inertial>
 struct GaugeH;
 template <size_t Dim, typename Frame = Frame::Inertial>
 struct SpacetimeDerivGaugeH;
+
+template<size_t Dim, typename>
```
Comment (nilsdeppe) [inline src/Evolution/Systems/GeneralizedHarmonic/TagsDeclarations.hpp:25]: Could you please name the second parameter?

File: src/Evolution/Systems/GeneralizedHarmonic/Characteristics.hpp [hunk 9]
```diff
@@ -0,0 +1,66 @@
...
+template <typename Tag>
+struct Normalized;
+}  // namespace Tags
+/// \endcond
+
+// IWYU pragma: no_forward_declare Tensor
+
+namespace GeneralizedHarmonic {
+/*!
+ * \ingroup GeneralizedHarmonic
```
Comment (nilsdeppe) [inline src/Evolution/Systems/GeneralizedHarmonic/Characteristics.hpp:27]: This group doesn't exist. We probably should have the GH namespace be in the GR group.

File: src/Evolution/Systems/GeneralizedHarmonic/Characteristics.hpp [hunk 10]
```diff
@@ -0,0 +1,66 @@
...
+ *
+ * The speeds \f$v\f$, are given by:
+ * \f{align*}
+ * v_{\psi} =& -(1 + \gamma_1) n_k N^k \\
+ * v_{0} =& -n_k N^k \\
+ * v_{\pm} =& -n_k N^k \pm N
+ * \f}
+ *
+ * where \f$N, N^k\f$ are the lapse and shift respectively, \f$\gamma_1\f$ is a
+ * constraint damping parameter, and \f$n_k\f$ is the unit normal to the
```
Comment (nilsdeppe) [inline src/Evolution/Systems/GeneralizedHarmonic/Characteristics.hpp:49]: We still seem to be using `n` as the normal for at least 2 different things, maybe 3. In this context we can use `t_k`

File: tests/Unit/Evolution/Systems/GeneralizedHarmonic/Test_Characteristics.cpp (hunk 6 above)
Comment (nilsdeppe) [reply tests/Unit/Evolution/Systems/GeneralizedHarmonic/Test_Characteristics.cpp:59]: As @wthrowe has pointed out, we've had a lot of bugs because people make the same mistakes twice...

**Review by nilsdeppe — DISMISSED** [review]
I'm fine with squashing changes.

Not for this PR, but we should generally decide on whether we want to use "ADM names" (N, N^i) or the  alpha, beta^i names, then be consistent throughout the code. Personally, I'm in favor of N, N^i, psi, g_{ij}, etc. that SpEC uses.

File: src/DataStructures/DataBox/Prefixes.hpp [hunk 11]
```diff
@@ -100,4 +100,16 @@ struct Next : db::PrefixTag, db::SimpleTag {
   using tag = Tag;
   static std::string name() noexcept { return "Next(" + Tag::name() + ")"; }
 };
+
+/// \ingroup DataBoxTagsGroup
+/// \brief Prefix corresponding to the characteristic speed of the field given
```
Comment (moxcodes) [inline src/DataStructures/DataBox/Prefixes.hpp:105]: might be helpful to say specifically `the characteristic speed of the characteristic field`

File: src/Evolution/Systems/GeneralizedHarmonic/Tags.hpp [hunk 12]
```diff
@@ -61,5 +62,42 @@ struct SpacetimeDerivGaugeH : db::SimpleTag {
   using type = tnsr::ab<DataVector, Dim, Frame>;
   static std::string name() noexcept { return "SpacetimeDerivGaugeH"; }
 };
+
+// {@
+/// \ingroup GeneralizedHarmonicGroup
+/// \brief Tags corresponding to the characteristic fields of the generalized
```
Comment (moxcodes) [inline src/Evolution/Systems/GeneralizedHarmonic/Tags.hpp:68]: `generalized harmonic system`

File: src/Evolution/Systems/GeneralizedHarmonic/Tags.hpp [hunk 13]
```diff
@@ -61,5 +62,42 @@ struct SpacetimeDerivGaugeH : db::SimpleTag {
...
+template<size_t Dim, typename Frame>
+struct UMinus {
+  using type = tnsr::aa<DataVector, Dim, Frame>;
+  static std::string name() noexcept { return "UMinus"; }
+};
+// @}
+
+template<size_t Dim, typename Frame>
+struct CharacteristicSpeeds : db::SimpleTag {
+  using type = Variables<db::wrap_tags_in<
```
Comment (moxcodes) [inline src/Evolution/Systems/GeneralizedHarmonic/Tags.hpp:97]: The characteristic speeds in `ValenciaDivClean` have type `std::array<DataVector, number_of_speeds>`. why is it different here? can we make them consistent? We would prefer `Variables` as is here.

File: src/Evolution/Systems/GeneralizedHarmonic/Characteristics.cpp [hunk 14]
```diff
@@ -0,0 +1,57 @@
...
+
+#define INSTANTIATION(_, data)                                      \
+  template struct GeneralizedHarmonic::CharacteristicSpeedsCompute< \
+      DIM(data), FRAME(data)>;
+
+GENERATE_INSTANTIATIONS(INSTANTIATION, (1, 2, 3),
+                        (Frame::Inertial, Frame::Grid))
+
+#undef INSTANTIATION
+#undef DIM
```
Comment (moxcodes) [inline src/Evolution/Systems/GeneralizedHarmonic/Characteristics.cpp:57]: `#undef FRAME`

File: src/Evolution/Systems/GeneralizedHarmonic/Characteristics.hpp (hunk 10 above)
Comment (moxcodes) [reply src/Evolution/Systems/GeneralizedHarmonic/Characteristics.hpp:49]: `n_k` is a spatial normal vector to an element face or an external boundary, so `t_k` would be inappropriate because it usually refers to a timelike normal to a hypersurface.

File: src/Evolution/Systems/GeneralizedHarmonic/TagsDeclarations.hpp (hunk 8 above)
Comment (moxcodes) [reply src/Evolution/Systems/GeneralizedHarmonic/TagsDeclarations.hpp:25]: clang-format (here and elsewhere)

File: tests/Unit/Evolution/Systems/GeneralizedHarmonic/TestFunctions.py [hunk 15]
```diff
@@ -43,3 +44,20 @@ def gauge_constraint(gauge_function, spacetime_normal_one_form,
...
+def char_speed_uzero(gamma1, lapse, shift, unit_normal):
+    return - np.dot(shift, unit_normal)
+
+
+def char_speed_uplus(gamma1, lapse, shift, unit_normal):
+    return - np.dot(shift, unit_normal) + lapse
+
+
+def char_speed_uminus(gamma1, lapse, shift, unit_normal):
+    return - np.dot(shift, unit_normal) - lapse
```
Comment (moxcodes) [inline tests/Unit/Evolution/Systems/GeneralizedHarmonic/TestFunctions.py:63]: `# End test functions for characteristics`

File: tests/Unit/Pypp/PyppFundamentals.hpp [hunk 16]
```diff
@@ -221,7 +221,8 @@ struct FromPyObject<long, std::nullptr_t> {
 #if PY_MAJOR_VERSION == 2 && PY_MINOR_VERSION == 7
     } else if (not PyInt_Check(t) and not PyLong_Check(t)) {
 #elif PY_MAJOR_VERSION == 3
-    } else if (not PyLong_Check(t)) {
+      // clang-tidy: hicpp-signed-bitwise
+    } else if (not PyLong_Check(t)) { // NOLINT
```
Comment (moxcodes) [inline tests/Unit/Pypp/PyppFundamentals.hpp:225]: why not use `PyLong_Check(t) == 0`?

**Review by moxcodes — CHANGES_REQUESTED** [review]
group code review with @markscheel @geoffrey4444 @fmahebert and Rob.

File: src/Evolution/Systems/GeneralizedHarmonic/Tags.hpp (hunk 13 above)
Comment (kidder) [reply src/Evolution/Systems/GeneralizedHarmonic/Tags.hpp:97]: Because doing so for Valencia was not possible prior to the previous commit...

Comment (kidder) [reply src/Evolution/Systems/GeneralizedHarmonic/Tags.hpp:97]: Also the characteristic fields are not currently implemented for Valencia

Comment (kidder) [reply src/Evolution/Systems/GeneralizedHarmonic/Tags.hpp:97]: I agree a `Variables` is preferable and we should make an issue to update the other systems as the characteristic fields are impemented

File: tests/Unit/Pypp/PyppFundamentals.hpp (hunk 16 above)
Comment (osheamonn) [reply tests/Unit/Pypp/PyppFundamentals.hpp:225]: Why change it? I was just trying to silence a clang-tidy warning and it would have to be done for every other check in this file.

File: src/Evolution/Systems/GeneralizedHarmonic/Tags.hpp (hunk 13 above)
Comment (osheamonn) [reply src/Evolution/Systems/GeneralizedHarmonic/Tags.hpp:97]: The reasoning was you can only defined characteristic fields for symmetric hyperbolic systems, 
and so for these systems the characteristics are thought about as "a set of fields each of which has a corresponding speed", whereas for hydro, the characteristics are matrices of left and right eigenvectors, and the speeds are the eigenvalues. The variables makes sense here because the fields and speeds are named and correspond to tags well. For implementing upwind flux it will then be useful to loop over the tags and get the corresponding speed and field. 

As to whether ValenciaDivClean should be changed, I think an array is perfectly fine to represent the speeds, but that doesn't seem like something this PR should be concerned with.

File: src/Evolution/Systems/GeneralizedHarmonic/Characteristics.hpp (hunk 10 above)
Comment (osheamonn) [reply src/Evolution/Systems/GeneralizedHarmonic/Characteristics.hpp:49]: I had thought that `n` was pretty consistently used to denote a spatial normal one-form. I also interpret `t` as being a spacetime normal to a hypersurface.

File: tests/Unit/Pypp/PyppFundamentals.hpp (hunk 16 above)
Comment (moxcodes) [reply tests/Unit/Pypp/PyppFundamentals.hpp:225]: Our thinking was just that treating the int return type more transparently is marginally better code style than needing to silence the linter. If it's a decision that would require numerous changes I am not personally that concerned about it (though the other Caltech reviewers should feel free to disagree).

Comment (fmahebert) [reply tests/Unit/Pypp/PyppFundamentals.hpp:225]: (I do not disagree)

File: src/Evolution/Systems/GeneralizedHarmonic/Tags.hpp (hunk 13 above)
Comment (hrueter) [reply src/Evolution/Systems/GeneralizedHarmonic/Tags.hpp:97]: @osheamonn I don't see what symmetric hyperbolicity has to do with that. I think that strong hyperbolicity should always suffice to define characteristic fields.

Comment (moxcodes) [reply src/Evolution/Systems/GeneralizedHarmonic/Tags.hpp:97]: We still think that Valencia should be changed to a Variables when able, but we agree with @osheamonn that it shouldn't be in this PR.

File: src/Evolution/Systems/GeneralizedHarmonic/Characteristics.hpp (hunk 10 above)
Comment (nilsdeppe) [reply src/Evolution/Systems/GeneralizedHarmonic/Characteristics.hpp:49]: ah yep, I was mistaken, sorry

**Comment by geoffrey4444** [comment]
Since #1129 supersede this, should we just close this one (#1071)?

**Comment by osheamonn** [comment]
Yes this can be closed.
